File: packages/wasmvm/wasmlib/ts/wasmclient/clientfunc.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmclient from "./index";
import { IKeyPair } from "./crypto";

export class ClientFunc {
    protected svc: wasmclient.Service;
    private keyPair: IKeyPair | null = null;
    private onLedger: boolean = false;
    private xfer: wasmclient.Transfer = new wasmclient.Transfer();

    constructor(svc: wasmclient.Service) {
        this.svc = svc;
    }


    // onLedger can override the default off-ledger to on-ledger posting
    public onLedgerRequest(onLedger: boolean): void {
        this.onLedger = onLedger;
    }

    //TODO rename to protected postRequest to prevent need for override and hide it

    // Sends a request to the smart contract service
    // You can wait for the request to complete by using the returned RequestID
    // as parameter to Service.waitRequest()
    public async post(hFuncName: wasmclient.Hname, args: wasmclient.Arguments | null): Promise<wasmclient.RequestID> {
        if (!args) args = new wasmclient.Arguments();

        if (!this.keyPair) this.keyPair = this.svc.keyPair;

        if (!this.keyPair) throw new Error("Key pair not defined");

        return await this.svc.postRequest(hFuncName, args, this.xfer, this.keyPair, this.onLedger);
    }

    // Optionally overrides the default keypair from the service
    public sign(keyPair: IKeyPair): void {
        this.keyPair = keyPair;
    }

    // Optionally indicates which tokens to transfer as part of the request
    // The tokens are presumed to be available in the signing account on the chain
    public transfer(xfer: wasmclient.Transfer): void {
        this.xfer = xfer;
    }
}

File: packages/wasmvm/wasmlib/go/wasmclient/clientfunc.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmclient

import (
	"github.com/iotaledger/wasp/packages/cryptolib"
)

type ClientFunc struct {
	svc      *Service
	keyPair  *cryptolib.KeyPair
	onLedger bool
	xfer     *Transfer
}

// OnLedgerRequest can override the default off-ledger to on-ledger posting
func (f *ClientFunc) OnLedgerRequest(onLedger bool) {
	f.onLedger = onLedger
}

// Post sends a request to the smart contract service
// You can wait for the request to complete by using the returned Request
// as parameter to Service.WaitRequest()
func (f *ClientFunc) Post(hFuncName uint32, args *Arguments) Request {
	keyPair := f.keyPair
	if keyPair == nil {
		keyPair = f.svc.keyPair
	}
	if args == nil {
		args = &Arguments{}
	}
	return f.svc.PostRequest(hFuncName, args.args, f.xfer, keyPair, f.onLedger)
}

// Sign optionally overrides the default keypair from the service
func (f *ClientFunc) Sign(keyPair *cryptolib.KeyPair) {
	f.keyPair = keyPair
}

// Transfer optionally indicates which tokens to transfer as part of the request
// The tokens are presumed to be available in the signing account on the chain
func (f *ClientFunc) Transfer(xfer *Transfer) {
	f.xfer = xfer
}
